# S.1 Numbers in Python

If you are students in Accounting, you should notice that floating point numbers are not reliable for keeping books. 
The simple code below would suffice to know the point.

In [5]:
s = 0
for i in range(10):
    s += 0.1
s

0.9999999999999999

The code does the following. 

- Set `s` to be zero.
- Repeat following for 10 times
  - reset `s` with old value of `s` plus 0.1
- Show s

You might have expected that the result was 1.0. But the output is different. 

When computers store a decimal numbers such as 0.1, it converts the number to have binary represenation. While all  integers have exact binary represenations, not all decimal numbers does. It's safe to understand all the decimal-looking numbers are approximations. In macroeconomic simulations, it doesn't give you much trouble.

This would cause, however, a fatal problem in some situation. Imagine a situation where I have paid you 0.1 billion Yen per year, for ten years. Your ledger is coded using float numbers. It records your 0.9999999999999999 billion receipt. Would you be angry as a creditor since I faild to pay you enough? Would you foreclose on my mortgage?

Chances are, programers in your company's engineering section would be fired for making a troublesome ledger. How would you avoid that misery?

Use `Decimal` when true decimal numbers are required. Otherwise, float is fine.

In [11]:
import decimal

d = decimal.Decimal('0.1')
s = 0
for i in range(10):
    s += d
s

Decimal('1.0')